bsmetadata/preprocessing_tools/html_parser/filters_and_cleaners.py

```diff
@@ -0,0 +1,459 @@
+        tags_to_remove_alone: Optional[List[TagToRemove]] = None,
+        attrs_to_keep: Optional[List[str]] = None,
+        start_parsing_at_tag: Optional[str] = "body",
+        consecutive_tags_to_fold: Optional[List[str]] = None,
+        convert_br_tag_to_breaking_line: Optional[bool] = False,
+        txt_max_chr_len_alone: float = -float("inf"),
+        txt_min_chr_len_alone: float = -float("inf"),
+        tags_exceptions_to_txt_max_min_chr_len_alone: List[str] = None,
+        txt_max_chr_len_with_content: float = -float("inf"),
+        txt_min_chr_len_with_content: float = -float("inf"),
+        tags_exceptions_to_txt_max_min_chr_len_with_content: List[str] = None,
+    ):
+        self.html_str = html_str
+        self.tags_to_remove_with_content = tags_to_remove_with_content
+        self.tags_to_remove_alone = tags_to_remove_alone
+        self.attrs_to_keep = attrs_to_keep
+        self.start_parsing_at_tag = start_parsing_at_tag
+        self.convert_br_tag_to_breaking_line = convert_br_tag_to_breaking_line
+
+        self.tags_to_remove_alone = (
+            [
+                TagToRemove(FAKE_TAG_BLOCK),
+                TagToRemove(FAKE_TAG_INLINE),
+                TagToRemove(FAKE_TAG_BASIC),
+            ]
+            if self.tags_to_remove_alone is None
+            else self.tags_to_remove_alone
+            + [
+                TagToRemove(FAKE_TAG_BLOCK),
+                TagToRemove(FAKE_TAG_INLINE),
+                TagToRemove(FAKE_TAG_BASIC),
+            ]
+        )
+
+        self.block_elements = BLOCK_ELEMENTS.copy()
+        if self.convert_br_tag_to_breaking_line:
+            self.block_elements.remove("br")
+            self.tags_to_remove_alone.append(TagToRemove("br"))
+
+        self.consecutive_tag_cleaner = ConsecutiveTagCleaner(
+            block_elements=self.block_elements,
+            consecutive_tags_to_fold=consecutive_tags_to_fold,
+        )
+
+        self.attribute_cleaner = AttributeCleaner(attrs_to_keep=attrs_to_keep)
+        self.tag_filter = TagFilter(
+            txt_max_chr_len_alone=txt_max_chr_len_alone,
+            txt_min_chr_len_alone=txt_min_chr_len_alone,
+            tags_exceptions_alone=tags_exceptions_to_txt_max_min_chr_len_alone,
+            txt_max_chr_len_with_content=txt_max_chr_len_with_content,
+            txt_min_chr_len_with_content=txt_min_chr_len_with_content,
+            tags_exceptions_with_content=tags_exceptions_to_txt_max_min_chr_len_with_content,
+            tags_to_remove_alone=self.tags_to_remove_alone,
+            tags_to_remove_with_content=tags_to_remove_with_content,
+        )
+
+    def apply(self):
+        html_str = self.html_str
+        # Traitement n°1: start the parsing at a special tags (mostly tested with <body>)
+        if self.start_parsing_at_tag is not None:
+            root = fromstring(html_str)
+            find = etree.XPath(f"//{self.start_parsing_at_tag}")
+            new_etree = find(root)[0]
```

Comment:
I guess this will throw an `IndexError` if the `self.start_parsing_at_tag` does not occur in the text (and I guess some html documents will not contain a `<body>` tag)? If so, it might make sense to wrap this error in some custom error class so that we can easily distinguish this "expected" error from unexpected ones.

Reply:
This is indeed an interesting question! I added your proposal, it now raises a `ValueError`. Is this ok?

Details about the body tag :
From what I understand, the body tag is part of the mandatory tags for an html code (see [this link](https://www.understandingcode.com/html-basics/required-tags/) for example). It turns out that the method used above `fromstring` will try to make the passed string legitimate and if ever it turns out that the body tag is missing, it will add it (see [this example](https://colab.research.google.com/drive/1A2YSS4HlCxBbjda4G9p8_yMLw9oEF_-i?usp=sharing) in a google colaboratory).

Reply:
I'm testing the code on tianjianjiang 's dataset and indeed several examples don't have a body tag and apparently the `fromstring` method doesn't add it. I'll investigate a bit more to see why this body tag is missing in some examples. 

If I need to make some changes for this new dataset, I will make them in another PR if it's ok for you.

Reply:
Sure, sounds great! 👍